homeassistant/components/proximity.py

```diff
@@ -0,0 +1,325 @@
+  devices:
+    - device_tracker.nwaring_nickmobile
+    - device_tracker.eleanorsiphone
+    - device_tracker.tsiphone
+  tolerance: 50
+"""
+
+import logging
+from homeassistant.helpers.event import track_state_change
+from homeassistant.helpers.entity import Entity
+from homeassistant.util.location import distance
+
+DEPENDENCIES = ['zone', 'device_tracker']
+
+# domain for the component
+DOMAIN = 'proximity'
+
+# default tolerance
+DEFAULT_TOLERANCE = 1
+
+# default zone
+DEFAULT_PROXIMITY_ZONE = 'home'
+
+# entity attributes
+ATTR_DIST_FROM = 'dist_to_zone'
+ATTR_DIR_OF_TRAVEL = 'dir_of_travel'
+ATTR_NEAREST = 'nearest'
+ATTR_FRIENDLY_NAME = 'friendly_name'
+
+# Shortcut for the logger
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup(hass, config):  # pylint: disable=too-many-locals,too-many-statements
+    """ get the zones and offsets from configuration.yaml"""
+    ignored_zones = []
+    if 'ignored_zones' in config[DOMAIN]:
+        for variable in config[DOMAIN]['ignored_zones']:
+            ignored_zones.append(variable)
+            _LOGGER.info('ignored zones loaded: %s', variable)
+
+    # get the devices from configuration.yaml
+    if 'devices' not in config[DOMAIN]:
+        _LOGGER.error('devices not found in config')
+        return False
+    else:
+        proximity_devices = []
+        for variable in config[DOMAIN]['devices']:
+            proximity_devices.append(variable)
+            _LOGGER.info('proximity device added: %s', variable)
+
+    # get the direction of travel tolerance from configuration.yaml
+    if 'tolerance' in config[DOMAIN]:
+        tolerance = config[DOMAIN]['tolerance']
+    else:
+        tolerance = DEFAULT_TOLERANCE
+    _LOGGER.debug('tolerance set to: %s', tolerance)
+
+    # get the zone to monitor proximity to from configuration.yaml
+    if 'zone' in config[DOMAIN]:
+        proximity_zone = config[DOMAIN]['zone']
+    else:
+        proximity_zone = DEFAULT_PROXIMITY_ZONE
+    _LOGGER.info('zone set to %s', proximity_zone)
+
+    entity_id = DOMAIN + '.' + proximity_zone
+    proximity_zone = 'zone.' + proximity_zone
+
+    state = hass.states.get(proximity_zone)
+    proximity_latitude = state.attributes.get('latitude')
+    proximity_longitude = state.attributes.get('longitude')
+    zone_friendly_name = state.attributes.get('friendly_name')
+
+    _LOGGER.debug('zone settings: LAT:%s LONG:%s', proximity_latitude,
+                  proximity_longitude)
+
+    # create an entity so that the proximity values can be used for other
+    # components
+    entities = set()
+
+    # set the default values
+    dist_to_zone = 'not set'
+    dir_of_travel = 'not set'
+    nearest = 'not set'
+
+    proximity = Proximity(hass, zone_friendly_name, dist_to_zone,
+                          dir_of_travel, nearest)
+    proximity.entity_id = entity_id
+
+    proximity.update_ha_state()
+    entities.add(proximity.entity_id)
+
+    def check_proximity_state_change(entity, old_state, new_state):
```

Comment:
So this method should be part of the `Proximity` class. Instead of updating attributes using `proximity.dist_to` you would use `self.dist_to = …`